### DISCUSSION TITLE
Kubernetes API discovery across namespaces

### DISCUSSION THREAD

**Discussion by felipeheld** [hazelcast/hazelcast#24684] [General]
When using Hazecast in its embedded topology and Hazelcast's Kubernetes's API for node discovery I believe it is not possible for nodes to discover one another across namespaces, given only one service name and one namespace definition are allowed in the discovery configuration.

The relevant code is here: https://github.com/hazelcast/hazelcast/blob/120d25873b173dbc3ec157b1d68c3980168978ef/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java#L243-L251

It would have been desirable in my case to have the endpoint be discoverable in more than one namespace, however I am not sure whether I'm actually right in my propositions nor if, were I right, this would be of actual use to other developers.

I'd be more than willing to create a PR, but I honestly believe it'd be better to just ask beforehand, just in case this is the intended behavior as deemed fit by the majority or by people more qualified in the project's development.

**Comment by SeriyBg** [comment]
Hello, @felipeheld! We do not support multi-namespace discovery. This is done on purpose since, usually, the discovery is via K8s services, which do not support multi-namespaces.

**Reply by felipeheld** [reply]
I see, thank you for your reply. I just wondered if it wouldn't be possible establishing multiple fully qualified domain names (service-name.namespace-name.svc) in the hazelcast.yml for the different services that would enable the discovery. It seems like a possibility, since we're only concerned with resolving those to their actual cluster internal ips.

But I'm throwing it out there, I'm not 100% on the correctness of my assumption.

**Reply by SeriyBg** [reply]
This is impossible, as only one `service-name` parameter is supported in the configuration. Could you please provide more details about your use case, in particular, to describe why the multi-namespace is needed and what problems you are trying to resolve? So that we can suggest a solution for it

**Reply by felipeheld** [reply]
Sorry for the lateness of my reply. I actually used Eureka for node discovery, so it eliminated the concern with multi-namespace discovery. But it just happens to have been an easier solution due to the fact we already used Eureka for our own services discovery.

I thought it would be an interesting addition if Eureka weren't the sole mean of discovery, in the event someone isn't using Eureka and doesn't want to set up a discovery server solely for hazelcast. But I guess that would be a very particular use case, especially since most people using Java in Kubernetes would likely be also using Eureka.

In any case, I see now there's probably no use for my suggestion, so this can be closed.

Thank you for your time in answering me!

**Reply by SeriyBg** [reply]
Good to hear that there is a workaround for your use case! Thank you for your input!
